src/main/java/org/embulk/input/marketo/delegate/ProgramMembersBulkExtractInputPlugin.java

```diff
@@ -0,0 +1,243 @@
+                java.util.function.Function<Set<String>, Iterable<ObjectNode>> getListIds = marketoService::getProgramsByIds;
+                programsToRequest = super.getObjectsByIds(idsStr, getListIds);
+            }
+            else {
+                programsToRequest = marketoService.getPrograms();
+            }
+            Iterator<ObjectNode> iterator = programsToRequest.iterator();
+            while (iterator.hasNext()) {
+                ObjectNode program = iterator.next();
+                int id = program.get("id").asInt();
+                if (!programIds.contains(id)) {
+                    programIds.add(id);
+                }
+            }
+            if (programIds.size() <= 0) {
+                throw new DataException("No program belong to this account.");
+            }
+            task.setExtractedPrograms(Optional.of(programIds));
+
+            ObjectNode result = marketoService.describeProgramMembers();
+            JsonNode fields = result.get("fields");
+            if (!fields.isArray()) {
+                throw new DataException("[fields] isn't array node.");
+            }
+            Map<String, String> extractFields = new HashMap<>();
+            for (JsonNode field : fields) {
+                String dataType = field.get("dataType").asText();
+                String name = field.get("name").asText();
+                if (!extractFields.containsKey(name)) {
+                    extractFields.put(name, dataType);
+                }
+            }
+            task.setProgramMemberFields(Optional.of(extractFields));
+        }
+    }
+
+    @Override
+    public ConfigDiff buildConfigDiff(PluginTask task, Schema schema, int taskCount, List<TaskReport> taskReports)
+    {
+        return CONFIG_MAPPER_FACTORY.newConfigDiff();
+    }
+
+    @Override
+    public TaskReport ingestServiceData(final PluginTask task, RecordImporter recordImporter, int taskIndex, PageBuilder pageBuilder)
+    {
+        TaskReport taskReport = CONFIG_MAPPER_FACTORY.newTaskReport();
+        ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(task.getNumberConcurrentExportJob());
+        if (Exec.isPreview()) {
+            return MarketoUtils.importMockPreviewData(pageBuilder, PREVIEW_RECORD_LIMIT);
+        }
+        else {
+            if (!task.getProgramMemberFields().isPresent() || !task.getExtractedPrograms().isPresent()) {
+                throw new ConfigException("program_member_fields or extracted_programs are missing.");
+            }
+            final MarketoRestClient restClient = createMarketoRestClient(task);
+            final List<String> fieldNames = new ArrayList<>(task.getProgramMemberFields().get().keySet());
+            final List<String> exportIDs = Collections.synchronizedList(new ArrayList<>());
+            CompletableFuture[] listFutureExportIDs = task.getExtractedPrograms().get().stream()
```

Comment:
`CompletableFuture` may not be a correct way to handle multiple threads/tasks here, as you can't interrupt running tasks if any of them throw an exception [See](https://stackoverflow.com/questions/29013831/how-to-interrupt-underlying-execution-of-completablefuture) => that means, it will wait for all other tasks to complete to cancel the current job => resources waited. 
I think, simple ThreadPoolExecutor & Runnable should work. `executor.execute(() -> {})`. When any thread throws exception, we can stop it with just `executor.shutdownNow()`